IceCap: Improve ring buffer stability

geky opened the pull request:
This includes a number of changes to improve the stability of the ring-buffers used to communicate between domains. As it was it was easy to lose notifications and cause applications to lock up.

The commits have more details, but here are the high-level changes:

1. RingBuffer is changed to enabled notifications by default, avoiding a race-condition during domain bringup.
2. Fixed an issue in BufferedRingBuffer where >ring-buffer sized messages could hang the ring-buffer if multiple messages are in the queue.
3. Reworked the virtio-console-server example a bit to use the BufferedRingBuffer more correctly.

This depends on https://gitlab.com/arm-research/security/icecap/icecap/-/merge_requests/22, which is depended on in several places in the commit history (and is already broken by a rebase needed there). So I will wait to fix the submodule and make this a non-draft until that is merged.

This should resolve https://github.com/veracruz-project/veracruz/issues/429, resolve https://github.com/veracruz-project/veracruz/issues/443, and fix the main hanging problem we are currently seeing in IceCap in CI.

mathias-arm commented:
I merged https://gitlab.com/arm-research/security/icecap/icecap/-/merge_requests/22 and rebased this PR.

geky commented:
Should be good to go now when/if CI passes

dominic-mulligan-arm commented:
@geky unfortunately some commits are not signed: can you rewrite the Git history to sign them?

dominic-mulligan-arm commented on workspaces/icecap-runtime/src/virtio-console-server/src/main.rs:
```diff
@@ -209,17 +193,18 @@ fn main(config: Config) -> Fallible<()> {
-            rb.rx_callback();
-            rb.tx_callback();
+        if badge & config.badges.rx != 0 {
+            //debug_println!("virtio-console-server: rx!");
```
Stray comment?

geky replied:
Will remove

dominic-mulligan-arm commented on runtime-manager/src/runtime_manager_icecap.rs:
```diff
@@ -52,75 +53,83 @@ fn main(config: Config) -> Fallible<()> {
     ) -> Self {
         Self {
-            channel: channel,
+            rb: BufferedRingBuffer::new(rb),
```
Is there a reason why `new` takes a `RingBuffer` instead of a `BufferedRingBuffer`?

geky replied:
`RingBuffer` is the underlying IceCap API we need to communicate over. `BufferedRingBuffer` is really a utility class that `RuntimeManager` uses to help manage the traffic on the ring-buffer.

Moving the `BufferedRingBuffer` out would just add an extra piece of work to the user of the API, and keeping it inside the `RuntimeManager` details allows it to change in the future without an API change.

At some point we may want to consider changing this to take a `Read + Write` type or something more generic, but at the moment it's not clear to me what that would be. `Read` and `Write` require std, and there is a `no_std` library for it, but if we want to use that or an IceCap specific trait is an open question.

dominic-mulligan-arm replied:
Ok great, thanks for clarifying.  I'll mark this as resolved.

mathias-arm commented:
You don't need to add commits to re-run CI, you can restart the failing jobs in "Actions".

geky commented:
> @geky unfortunately some commits are not signed: can you rewrite the Git history to sign them?

I was going to just remove those commits, they're empty and was just to trigger CI.

> You don't need to add commits to re-run CI, you can restart the failing jobs in "Actions".

Unfortunately this clobbers the log and hides the old results. I was particularly interested in how many runs failed (though not sure what to do with this info).

mathias-arm commented:
> > You don't need to add commits to re-run CI, you can restart the failing jobs in "Actions".
> 
> Unfortunately this clobbers the log and hides the old results. I was particularly interested in how many runs failed (though not sure what to do with this info).

You can access the logs for all attempts. e.g. https://github.com/veracruz-project/veracruz/actions/runs/2247521453 see the `Latest #2` drop-down allows you to select the attempt you want to look at.

dominic-mulligan-arm commented:
> > > You don't need to add commits to re-run CI, you can restart the failing jobs in "Actions".
> > 
> > 
> > Unfortunately this clobbers the log and hides the old results. I was particularly interested in how many runs failed (though not sure what to do with this info).
> 
> You can access the logs for all attempts. e.g. https://github.com/veracruz-project/veracruz/actions/runs/2247521453 see the `Latest #2` drop-down allows you to select the attempt you want to look at.

*The more you know!*

dominic-mulligan-arm commented:
@mathias-arm I went ahead and removed the unverified and useless commits added to force a CI re-run, and addressed my own comment above.  Can you review this (given it risks needing a rebase given the amount of other changes going in to the repository)?

geky commented:
> You can access the logs for all attempts. e.g. https://github.com/veracruz-project/veracruz/actions/runs/2247521453 see the Latest #2 drop-down allows you to select the attempt you want to look at.

Oh this is good to know! Looks like they must have added this recently.

> @mathias-arm I went ahead and removed the unverified and useless commits added to force a CI re-run, and addressed my own comment above.

Thanks for fixing it up. This looks good to me though not sure my review counts.

---

It should be noted we're still seeing unknown CI failures in IceCap, but this PR at least fixes some of the known reasons. More fixes can always come later.

dominic-mulligan-arm commented:
No problem, was a small change to get it over the line and remove any risk of rebase nightmares.

mathias-arm commented:
> It should be noted we're still seeing unknown CI failures in IceCap, but this PR at least fixes some of the known reasons. More fixes can always come later.

Since disabling the 3 tests and the `yield_()` work-around, I don't remember seeing spurious CI failures. The changes look good, but I would be uncomfortable to approve if this PR is going to make CI unstable again. Let's see the result of these checks and after a rebase. If both are successful we will merge.

mathias-arm reviewed (approved):
Approval in principle, conditioned to rebase and successful CI results.

dominic-mulligan-arm commented:
Valid concerns --- @geky can you do the rebase?

geky commented:
I'm not opposed to removing bad3596 for now, that would make this PR purely an improvement over main.

But I am concerned about the failures for f5087dc, since it in theory had all the fixes in it...

geky commented:
> Valid concerns --- @geky can you do the rebase?

Can do, but should we go ahead and remove https://github.com/veracruz-project/veracruz/commit/bad3596bc720cbd163cef02a8e1245d07a60d93e?

dominic-mulligan-arm commented:
Unverified commits 😢

mathias-arm commented:
bad3596 can be dealt separately in #416. But I don't want to assume it is the only contributor to CI instability. As @geky pointed out f5087dc (now 66d5d48) failed in CI.

geky commented:
> Unverified commits 😢

Woopsie doopsie, fixed now

dominic-mulligan-arm commented:
What is the status of this, again?  I've lost track of what's going on with this PR, unfortunately...

mathias-arm commented:
Superseded by #555
